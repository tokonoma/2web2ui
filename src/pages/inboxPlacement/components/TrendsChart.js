import React, { useEffect, useMemo } from 'react';
import { connect } from 'react-redux';
import { Bar, Rectangle } from 'recharts';
import moment from 'moment';
import { tokens } from '@sparkpost/design-tokens-hibana';

import BarChart from 'src/components/charts/BarChart';
import Legend from 'src/components/charts/Legend';
import { roundToPlaces } from 'src/helpers/units';
import TooltipMetric from 'src/components/charts/TooltipMetric';
import useDateHover from 'src/pages/inboxPlacement/hooks/useDateHover';
import { getInboxPlacementTrends } from 'src/actions/inboxPlacement';
import { selectTrends } from 'src/selectors/inboxPlacement';
import Callout from 'src/components/callout';
import { formatApiDate } from 'src/helpers/date';
import { PanelSectionLoading } from 'src/components/loading';

const yKeys = [
  {
    fill: tokens.color_blue_1000,
    key: 'missing',
    label: 'Missing',
  },
  {
    fill: tokens.color_blue_400,
    key: 'spam',
    label: 'Spam',
  },
  {
    fill: tokens.color_blue_600,
    key: 'inbox',
    label: 'Inbox',
  },
];

const legend = [
  ...yKeys.map(({ fill, label }) => ({ label, fill })).reverse(),
  { fill: tokens.color_white, label: 'No Tests Sent', hasBorder: true },
];

const panelHeight = '280px';

const yAxisProps = {
  tickFormatter: tick => `${roundToPlaces(tick * 100, 0)}%`,
  domain: [0, 1],
  ticks: [0, 0.25, 0.5, 0.75, 1.0],
};

const getTooltipContent = ({ payload = {} }) => (
  <>
    <TooltipMetric label={'Total Messages'} value={payload.totalMessages} />
    {yKeys
      .map(({ fill, label, key }) => (
        <TooltipMetric
          key={key}
          color={fill}
          label={label}
          value={`${(payload[key] * 100).toFixed(0)}%`}
        />
      ))
      .reverse()}
  </>
);

export const TrendsChart = props => {
  const [hoveredDate, handleDateHover, resetDateHover] = useDateHover();
  const { getInboxPlacementTrends, trends, hasNoData, loading, error, filters = {} } = props;

  useEffect(() => {
    const {
      dateRange: { from, to },
      tags,
    } = filters;

    getInboxPlacementTrends({
      from: formatApiDate(from),
      to: formatApiDate(to),
      ...Object.keys(tags).reduce((acc, curr) => {
        if (tags[curr]) {
          acc[curr] = tags[curr];
        }
        return acc;
      }, {}),
    });
  }, [getInboxPlacementTrends, filters]);

  const xAxisProps = useMemo(() => {
    const interval = Math.floor(trends.length / 10);
    const xTicks = trends.filter((value, index) => index % interval === 0).map(({ date }) => date);
    return {
      ticks: xTicks,
      tickFormatter: tick => moment(tick).format('M/D'),
    };
  }, [trends]);

  const barShape = props => {
    const isActiveDate = props.date === hoveredDate;
    const isOpaque = hoveredDate && !isActiveDate;
    return <Rectangle {...props} style={{ transition: '0s' }} opacity={isOpaque ? 0.5 : 1} />;
  };

  const renderBars = yKeys.map(({ key, fill }) => (
    <Bar
      stackId="stack"
      key={key}
      dataKey={key}
      onMouseOver={handleDateHover}
      fill={fill}
      isAnimationActive={false}
      minPointSize={0}
      cursor="pointer"
      shape={barShape}
    />
  ));

  if (error) {
    return (
      <div>
        <Callout title="Unable to Load Trends Data" height={panelHeight}>
          {error.message}
        </Callout>
      </div>
    );
  }

  if (loading) {
    return <PanelSectionLoading minHeight={panelHeight} />;
  }

  return (
    <>
      {hasNoData ? (
        <div>
          <Callout height={panelHeight}>Inbox Placement Trends Not Available</Callout>
        </div>
      ) : (
        <>
          {/*float:right does't work for some reason. Causes tooltip to not show when hovering bottom of chart*/}
          <div style={{ display: 'flex', justifyContent: 'flex-end' }}>
            <Legend items={legend} />
          </div>
          <div className="LiftTooltip" onMouseOut={resetDateHover}>
            <BarChart
              gap={1}
              hasBackgroundBars={false}
              hovered={hoveredDate}
              onMouseOver={handleDateHover}
              margin={{ top: 10, left: 30, right: 10, bottom: 25 }}
              tooltipContent={getTooltipContent}
              timeSeries={trends}
              tooltipWidth="150px"
              yAxisProps={yAxisProps}
              xAxisProps={xAxisProps}
            >
              {renderBars}
            </BarChart>
          </div>
        </>
      )}
    </>
  );
};

const mapStateToProps = (state, ownProps) => {
  const trends = selectTrends(state, ownProps);

  return {
    error: state.inboxPlacement.getTrendsError,
    loading: state.inboxPlacement.getTrendsPending,
    trends: trends,
    hasNoData: trends.length === 0,
  };
};

export default connect(mapStateToProps, { getInboxPlacementTrends })(TrendsChart);
