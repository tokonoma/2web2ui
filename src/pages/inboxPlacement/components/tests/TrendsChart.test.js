import { shallow } from 'enzyme';
import React from 'react';
import { TrendsChart } from '../TrendsChart';

describe('Inbox Placement Trends Chart', () => {
  const props = {
    hoveredDate: '',
    loading: false,
    error: null,
    hasNoData: false,
    trends: [
      {
        date: '2019-11-10',
        totalMessages: 10,
        inbox: 0.5,
        spam: 0.5,
        missing: 0,
      },
      {
        date: '2019-11-11',
        totalMessages: 40,
        inbox: 0.8,
        spam: 0.15,
        missing: 0.05,
      },
    ],
  };

  const subject = (options = {}) => shallow(<TrendsChart {...props} {...options} />);

  it('renders happy path correctly', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    const wrapper = subject({ loading: true });
    expect(wrapper.find('PanelSectionLoading')).toExist();
  });

  it('renders error correctly', () => {
    const wrapper = subject({ error: { message: 'You Dun Goofed' } });
    expect(wrapper.find('Callout')).toExist();
    expect(wrapper.find('Callout')).toHaveProp('title', 'Unable to Load Trends Data');
    expect(wrapper.find('Callout')).toHaveTextContent('You Dun Goofed');
  });

  it('renders no account level data callout correctly', () => {
    const wrapper = subject({ trends: [], hasNoData: true });
    expect(wrapper.find('Callout')).toExist();
    expect(wrapper.find('Callout')).toHaveTextContent('Inbox Placement Trends Not Available');
  });

  describe('Bars', () => {
    const getShapeFn = hoveredDate => {
      const wrapper = subject();
      const changeHovered = wrapper
        .find('Bar')
        .first()
        .prop('onMouseOver');
      changeHovered({ payload: { date: hoveredDate } });
      wrapper.update();

      return wrapper
        .find('Bar')
        .first()
        .prop('shape');
    };

    it('renders hovered bar correctly', () => {
      const shapeFn = getShapeFn('2019-11-10');
      const rectangleComponent = shapeFn({ date: '2019-11-10' });
      expect(rectangleComponent.props.opacity).toEqual(1);
    });

    it('renders un-hovered bar correctly when another bar is hovered', () => {
      const shapeFn = getShapeFn('2019-11-10');
      const rectangleComponent = shapeFn({ date: '2019-11-11' });
      expect(rectangleComponent.props.opacity).toEqual(0.5);
    });

    it('renders un-hovered bar correctly when no bar is hovered', () => {
      const shapeFn = getShapeFn('');
      const rectangleComponent = shapeFn({ date: '2019-11-10' });
      expect(rectangleComponent.props.opacity).toEqual(1);
    });
  });
});
