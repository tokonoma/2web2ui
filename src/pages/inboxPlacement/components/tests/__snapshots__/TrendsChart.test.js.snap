// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inbox Placement Trends Chart renders happy path correctly 1`] = `
<Fragment>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-end",
      }
    }
  >
    <Legend
      items={
        Array [
          Object {
            "fill": "#429bff",
            "label": "Inbox",
          },
          Object {
            "fill": "#abd2ff",
            "label": "Spam",
          },
          Object {
            "fill": "#203752",
            "label": "Missing",
          },
          Object {
            "fill": "#ffffff",
            "hasBorder": true,
            "label": "No Tests Sent",
          },
        ]
      }
    />
  </div>
  <div
    className="LiftTooltip"
    onMouseOut={[Function]}
  >
    <BarChart
      activeFill="#22838A"
      fill="#B3ECEF"
      gap={1}
      hasBackgroundBars={false}
      height={250}
      hovered=""
      margin={
        Object {
          "bottom": 25,
          "left": 30,
          "right": 10,
          "top": 10,
        }
      }
      onMouseOver={[Function]}
      shouldHighlightSelected={true}
      timeSeries={
        Array [
          Object {
            "date": "2019-11-10",
            "inbox": 0.5,
            "missing": 0,
            "spam": 0.5,
            "totalMessages": 10,
          },
          Object {
            "date": "2019-11-11",
            "inbox": 0.8,
            "missing": 0.05,
            "spam": 0.15,
            "totalMessages": 40,
          },
        ]
      }
      tooltipContent={[Function]}
      tooltipWidth="150px"
      width="99%"
      xAxisProps={
        Object {
          "tickFormatter": [Function],
          "ticks": Array [],
        }
      }
      xAxisRefLines={Array []}
      xKey="date"
      yAxisProps={
        Object {
          "domain": Array [
            0,
            1,
          ],
          "tickFormatter": [Function],
          "ticks": Array [
            0,
            0.25,
            0.5,
            0.75,
            1,
          ],
        }
      }
      yAxisRefLines={Array []}
      yKey="value"
      yRange={
        Array [
          "auto",
          "auto",
        ]
      }
    >
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        cursor="pointer"
        data={Array []}
        dataKey="missing"
        fill="#203752"
        hide={false}
        isAnimationActive={false}
        key="missing"
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        onMouseOver={[Function]}
        shape={[Function]}
        stackId="stack"
        xAxisId={0}
        yAxisId={0}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        cursor="pointer"
        data={Array []}
        dataKey="spam"
        fill="#abd2ff"
        hide={false}
        isAnimationActive={false}
        key="spam"
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        onMouseOver={[Function]}
        shape={[Function]}
        stackId="stack"
        xAxisId={0}
        yAxisId={0}
      />
      <Bar
        animationBegin={0}
        animationDuration={400}
        animationEasing="ease"
        cursor="pointer"
        data={Array []}
        dataKey="inbox"
        fill="#429bff"
        hide={false}
        isAnimationActive={false}
        key="inbox"
        layout="vertical"
        legendType="rect"
        minPointSize={0}
        onAnimationEnd={[Function]}
        onAnimationStart={[Function]}
        onMouseOver={[Function]}
        shape={[Function]}
        stackId="stack"
        xAxisId={0}
        yAxisId={0}
      />
    </BarChart>
  </div>
</Fragment>
`;
